### PR TITLE
Feat: Style customisation in the button widget

### LINKED ISSUE #4149: [Feature] Style customisation in the button widget.
## Summary

As a user, I should be able to change button style as follows:
 - The background colour can change with a hex code
 - font colour can change with hex code
 - elevation (box-shadow)
 - There should be button variant `contained (solid button)`, `outlined (only borders)`, `text (text buttons)`
 - Button can be multiple sizes: Large, Medium, Small
 - Button can have an `end icon` or `start icon`
 - ~~Button `end/start icon` can have a dropdown group as well. (button adornment)~~

These changes should also extend to the button widget used elsewhere, Ex: in the table widget.

## Motivation

I should be able to customize the button beyond primary/secondary/danger types.

## RACI matrix

| Attributes | Members | 
| ------------- | ------------- |
| Responsible |  @wmdev0808  |
| Accountable | @somangshu  |
| Consulted | @areyabhishek, @momcilo-appsmith  |
| Informed |  @Nikhil-Nandagopal |

### PR DESCRIPTION
**Pull request by wmdev0808**
## Description

- Add relevant properties for style customization
- Complete the first MVP of the button widget with the newly added style properties

Fixes #4149 

## Type of change
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

## How Has This Been Tested?

- Test A
- Test B

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: feature/4149-button-widget-style 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 55.08 **(-0.04)** | 37.05 **(-0.21)** | 33.72 **(-0.06)** | 55.67 **(-0.04)**
 :red_circle: | app/client/src/components/designSystems/appsmith/IconButtonComponent.tsx | 50 **(-9.18)** | 8.33 **(-3.67)** | 22.22 **(-14.14)** | 60.61 **(-8.44)**
 :red_circle: | app/client/src/components/designSystems/blueprint/ButtonComponent.tsx | 48.39 **(-8.87)** | 23.04 **(-27.46)** | 38.46 **(-17.1)** | 47.32 **(-10.62)**
 :green_circle: | app/client/src/components/editorComponents/DropdownComponent.tsx | 23.88 **(0.69)** | 14.81 **(0.52)** | 6.67 **(0)** | 27.27 **(3.08)**
 :green_circle: | app/client/src/components/propertyControls/BorderRadiusOptionsControl.tsx | 58.06 **(-1.94)** | 33.33 **(-6.67)** | 33.33 **(11.11)** | 62.96 **(2.09)**
 :green_circle: | app/client/src/mockResponses/WidgetConfigResponse.tsx | 14.02 **(1.64)** | 0 **(0)** | 0 **(0)** | 15 **(1.73)**
 :green_circle: | app/client/src/utils/autocomplete/TernServer.ts | 50.71 **(0.24)** | 38.6 **(0.88)** | 36.21 **(0)** | 55.35 **(0.27)**
 :red_circle: | app/client/src/widgets/ButtonWidget.tsx | 64 **(-11)** | 37.5 **(-12.5)** | 50 **(0)** | 63.27 **(-11.02)**
 :green_circle: | app/client/src/widgets/IconButtonWidget.tsx | 50 **(1.52)** | 0 **(0)** | 40 **(0)** | 50 **(1.61)**</details>

### CODE REVIEW

**Review by somangshu — DISMISSED** [review]
Approving to run tests

**Comment by somangshu** [comment]
Rerunning tests because of unexpected failures

**Comment by somangshu** [comment]
@wmdev0808 feedback from the deploy preview.
- The border radius in the rounded corner mode should be reduced to 5px.
- The automated dark font colour is right now `rgb(169, 167, 167) !important`, should change to `#333`.
- Icon colour should inherit the colour from parent text, Icon colour prop should be removed
- Icon alignment should be should be set to default when an icon is set for the first time.
- In the icon control `none` should always be the first selection as there is no other way to remove the icon. This will have a global effect
- Reordering sections: 
    - General
        - Label, Visible, Disable, Google recaptcha key, Google recaptcha V2
    - Actions (All properties)
    - Style (The rest of the style attributes)

**Review by somangshu — DISMISSED** [review]
Approving to run tests.

@wmdev0808 please check the pending comments

**Review by somangshu — APPROVED** [review]
LGTM

File: app/client/src/constants/DefaultTheme.tsx [hunk 1]
```diff
@@ -567,6 +567,147 @@ type ColorType = {
...
+    };
+    disabled: {
+      bgColor: Color;
+      textColor: Color;
+    };
+    /**
+     * PRIMARY style
+     */
+    primary: {
+      solid: {
```
Comment (SatishGandham) [inline app/client/src/constants/DefaultTheme.tsx:587]: Can you extra button properties to a separate type and use it.

**Review by SatishGandham — CHANGES_REQUESTED** [review]

**Review by SatishGandham — DISMISSED** [review]

**Comment by somangshu** [comment]
Rerunning all tests, Many unrelated tests are failing

**Comment by somangshu** [comment]
@wmdev0808 please merge latest release, resolve conflicts and push back to done

**Review by somangshu — DISMISSED** [review]

**Review by somangshu — APPROVED** [review]

**Comment by somangshu** [comment]
Rerunning all jobs, Everything failed in the last run

**Review by somangshu — DISMISSED** [review]

**Comment by somangshu** [comment]
@arslanhaiderbuttar can you drop an update here

**Comment by somangshu** [comment]
@vivekverma2312 I see some functionalities not working here any more, Can you verify all the [requested changes](https://github.com/appsmithorg/appsmith/pull/6052#issuecomment-885438739) again. Moving this back to QA pipeline.

**Comment by vivekverma2312** [comment]
@wmdev0808 On selecting custom in Button style the Background change property is not coming. Basically there is no change if user selects custom button style.
Screenshot 2021-08-18 at 11 19 25 AM

**Comment by vivekverma2312** [comment]
@wmdev0808 Oops something went wrong on widget and getting `Cannot read property 'hoverColor' of undefined` in console error on button variant JS property.
https://www.loom.com/share/47e67a41a7d74e9b85e7ca27d80a8edd

**Review by somangshu — APPROVED** [review]
Approving for tests to run

**Comment by somangshu** [comment]
@vivekverma2312 we need to recheck if the widget has the missing functionalities now

**Comment by vivekverma2312** [comment]
@wmdev0808 Deploy preview is tested and all the bugs are fixed. Good to go to release.

**Review by somangshu — APPROVED** [review]
Approving to run tests

**Comment by somangshu** [comment]
Rerunning all tests here, `LoginFromUIApp_spec.js` failed

**Review by somangshu — APPROVED** [review]
LGTM, Changes related to the security issue where the package.json and yarn.json had a diff.

**Pull request merged**